### PR TITLE
TRPL: embedding rust

### PR DESCRIPTION
**Pull request by steveklabnik**
This is the first chunk of the last guide I'll land before 1.0. I wanted to kick off reviews earlier rahter than later, so here's the important parts, with the Python and Node implementations coming later today.

### CODE REVIEW

**Comment by steveklabnik** [comment]
This is based off of work by both @alexcrichton and @callahad. :heart: 

r? @alexcrichton

File: src/doc/trpl/rust-inside-other-languages.md [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+                             ^~~~~
+```
+
+That first warning is because we are building a library. If we had a test
+for this function, the warning would go away. But for now, it’s never
+called.
+
+The second is related to `x` versus `_x`. Because we never actually _do_
+anything with `x`, we get a warning about it. In our case, that’s perfectly
+okay, as we’re just trying to waste CPU cycles. Prefixing `x` with the
```
Comment (Nercury) [inline src/doc/trpl/rust-inside-other-languages.md]: Unless optimiser optimises this away?

File: src/doc/trpl/rust-inside-other-languages.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+want this to be as fast as possible! You can find the output of the library in
+`target/release`:
+
+```bash
+$ ls target/release/
+build  deps  examples  libembed.so  native
+```
+
+That `libembed.so` is our ‘shared object’ library. We can use this file
+just like any shared object library written in C!
```
Comment (callahad) [inline src/doc/trpl/rust-inside-other-languages.md]: Should we mention that you might also see `.dylib` or `.dll` depending on your host platform?

File: src/doc/trpl/rust-inside-other-languages.md (hunk 1 above)
Comment (callahad) [inline src/doc/trpl/rust-inside-other-languages.md]: Yeah, I had to be careful to use a debug build of my "spin uselessly" library, lest nothing happened at all. :)

**Comment by steveklabnik** [comment]
Added python and node. So easy!

File: src/doc/trpl/rust-inside-other-languages.md (hunk 2 above)
Comment (steveklabnik) [inline src/doc/trpl/rust-inside-other-languages.md]: :+1:

File: src/doc/trpl/rust-inside-other-languages.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/trpl/rust-inside-other-languages.md]: I added a `let _ = x` line, and performance didn't seem to change. is the optimizer smart enough in these cases?

File: src/doc/trpl/rust-inside-other-languages.md [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+        thread::spawn(|| {
+            let mut _x = 0;
+            for _ in (0..5_000_001) {
+                _x += 1
+            }
+        })
+    }).collect();
+
+    for h in handles {
+        h.join().ok().expect(“Could not join a thread!”);
```
Comment (killercup) [inline src/doc/trpl/rust-inside-other-languages.md]: s/“/"
s/”/"

File: src/doc/trpl/rust-inside-other-languages.md [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+though. The first is modify the beginning of our code:
+
+```rust,ignore
+#[no_mangle]
+pub extern fn process() {
+```
+
+We have to add a new attribute, `no_mangle`. When you create a Rust library, it
+changes the name of the function in the compiled output. The reasons that this
+are done are outside the scope of this tutorial, but in order for other
```
Comment (killercup) [inline src/doc/trpl/rust-inside-other-languages.md]: s/that this are/that this is

or just "The reasons for this are outside the scope…"

File: src/doc/trpl/rust-inside-other-languages.md (hunk 3 above)
Comment (steveklabnik) [inline src/doc/trpl/rust-inside-other-languages.md]: oh dammit vim

**Comment by steveklabnik** [comment]
more nits addressed. r? @alexcrichton

File: src/doc/trpl/rust-inside-other-languages.md [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+`target/release`:
+
+```bash
+$ ls target/release/
+build  deps  examples  libembed.so  native
+```
+
+That `libembed.so` is our ‘shared object’ library. We can use this file
+just like any shared object library written in C! As an aside, this may be
+`libembed.dll` or `libembed.dylib`, depending on the platform.
```
Comment (alexcrichton) [inline src/doc/trpl/rust-inside-other-languages.md]: s/libembed.dll/embed.dll/

File: src/doc/trpl/rust-inside-other-languages.md [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+
+The second is related to `x` versus `_x`. Because we never actually _do_
+anything with `x`, we get a warning about it. In our case, that’s perfectly
+okay, as we’re just trying to waste CPU cycles. Prefixing `x` with the
+underscore removes the warning.
+
+Finally, we join on each thread.
+
+Comparing this code to the Ruby implementation, you’ll notice that it’s very
+similar looking, but has more noise. Ahh, language design tradeoffs. 😊
```
Comment (alexcrichton) [inline src/doc/trpl/rust-inside-other-languages.md]: This phrase seems a bit odd (to call out that Rust is more noisy). I would personally consider the choice to use `.map` + `.collect` instead of `Vec::new` + `.push` to be adding a bit of "noise" here and otherwise the only other part seems to be the `ok().expect()` which I believe is more commonly used as `.unwrap()` which doesn't seem really that much more noisy.

Overall I think it's fine to point out the similarities here, but we may not want to call out much about design tradeoffs here.

**Comment by alexcrichton** [comment]
Just a few minor nits here and there, but otherwise r=me, thanks @steveklabnik!

File: src/doc/trpl/rust-inside-other-languages.md (hunk 5 above)
Comment (steveklabnik) [inline src/doc/trpl/rust-inside-other-languages.md]: ah nice

File: src/doc/trpl/rust-inside-other-languages.md (hunk 6 above)
Comment (steveklabnik) [inline src/doc/trpl/rust-inside-other-languages.md]: I think I was just trying to be overly charitable that this isn't about "Ruby sucks, rust is great!" and to acknowledge our weaknesses. I'll just remove it.
